### DISCUSSION TITLE
Docker containers and service discovery

### DISCUSSION THREAD

**Discussion by Waste-Treatment** [microsoft/aspire#5862] [Q&A] [answered]
We may be missing the intent of the relationship between Aspire and containers but here goes. 

This scenario was us trying weave in our function app using a container as Aspire does not support function apps.
However, I think the question is generally applicable to anything running in a container not directly config'd by one of the 
container extension methods.

In AppHost, I have this setup and the container starts up just fine.

```csharp 
var containerService = builder.AddDockerfile(
    "SomeServiceInaContainer",
    "../ContextWhereDockerFileLives")
    .WithEndpoint(scheme: "http", targetPort: 80, port: 8080);
```

However, the following does not work...

```csharp
builder.AddProject<ABlazorClient>("client")
.WithReference(containerService)
.WithExternalHttpEndpoints();
```

... and produces

Argument 2: cannot convert from 'Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.ContainerResource>' to 'Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.IResourceWithConnectionString>'

Are we approaching this scenario correctly?
Are container services able to participate in service discovery?
If so, how should we approach this?

Thanks!

**Comment by davidfowl** [comment, marked as answer]
You need to explicitly reference an endpoint. When you are using a base container, there's no service discovery assumed (it could be a connection string you wanted injected vs service discovery information). 

```C#
builder.AddProject<ABlazorClient>("client")
.WithReference(containerService.GetEndpoint("http"))
.WithExternalHttpEndpoints();
```

See this [WithReference overload](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.resourcebuilderextensions.withreference?view=dotnet-aspire-8.0#aspire-hosting-resourcebuilderextensions-withreference-1(aspire-hosting-applicationmodel-iresourcebuilder((-0))-aspire-hosting-applicationmodel-endpointreference))

**Reply by Waste-Treatment** [reply]
Thank you for the quick response.
